src/SAWScript/Prover/MRSolver/Solver.hs

```diff
@@ -345,31 +349,65 @@ handling the recursive ones
+proveCoIndHypPreCond :: CoIndHyp -> MRM ()
+proveCoIndHypPreCond hyp =
+  do (pre1, pre2) <- applyCoIndHypPreconds hyp
+     pre <- liftSC2 scAnd pre1 pre2
+     success <- mrProvable pre
+     if success then return () else
+       throwMRFailure $
+       PrecondNotProvable (coIndHypLHSFun hyp) (coIndHypRHSFun hyp) pre
+
+-- | Co-inductively prove the refinement
+--
+-- > forall x1, ..., xn. preF y1 ... ym -> preG z1 ... zl ->
+-- >   F y1 ... ym |= G z1 ... zl@
+--
+-- where @F@ and @G@ are the given 'FunName's, @y1, ..., ym@ and @z1, ..., zl@
+-- are the given argument lists, @x1, ..., xn@ is the current context of uvars,
+-- and @preF@ and @preG@ are the preconditions associated with @F@ and @G@,
+-- respectively. This proof is performed by coinductively assuming the
+-- refinement holds and proving the refinement with the definitions of @F@ and
+-- @G@ unfolded to their bodies. Note that this refinement is performed with
+-- /only/ the preconditions @preF@ and @preG@ as assumptions; all other
+-- assumptions are thrown away. If while running the refinement computation a
+-- 'CoIndHypMismatchWidened' error is reached with the given names, the state is
+-- restored and the computation is re-run with the widened hypothesis.
+mrRefinesCoInd :: FunName -> [Term] -> FunName -> [Term] -> MRM ()
+mrRefinesCoInd f1 args1 f2 args2 =
   do ctx <- mrUVarCtx
-     withCoIndHyp' (CoIndHyp ctx f1 f2 args1 args2) m
-
--- | The main loop of 'withCoIndHyp'
-withCoIndHyp' :: CoIndHyp -> MRM a -> MRM a
-withCoIndHyp' hyp m =
-  withCoIndHypRaw hyp m `catchError` \case
-  MRExnWiden nm1' nm2' new_vars
-    | coIndHypLHSFun hyp == nm1' && coIndHypRHSFun hyp == nm2' ->
-        -- NOTE: the state automatically gets reset here because we defined MRM
-        -- with ExceptT at a lower level than StateT
-        do mrDebugPPPrefixSep 1 "Widening recursive assumption for" nm1' "|=" nm2'
-           debugPrint 2 $ "Widening indices: " ++ intercalate ", " (map show new_vars)
-           hyp' <- generalizeCoIndHyp hyp new_vars
-           withCoIndHyp' hyp' m
-  e -> throwError e
+     preF1 <- mrGetPrecond f1
+     preF2 <- mrGetPrecond f2
+     let hyp = CoIndHyp ctx f1 f2 args1 args2 preF1 preF2
+     proveCoIndHypPreCond hyp
+     proveCoIndHyp hyp
+
+-- | Prove the refinement represented by a 'CoIndHyp' coinductively. This is the
+-- main loop implementing 'mrRefinesCoInd'. See that function for documentation.
+proveCoIndHyp :: CoIndHyp -> MRM ()
+proveCoIndHyp hyp =
+  do let f1 = coIndHypLHSFun hyp
+         f2 = coIndHypRHSFun hyp
+         args1 = coIndHypLHS hyp
+         args2 = coIndHypRHS hyp
+     debugPretty 1 ("proveCoIndHyp" <+> ppInEmptyCtx hyp)
+     lhs <- fromMaybe (error "proveCoIndHyp") <$> mrFunBody f1 args1
+     rhs <- fromMaybe (error "proveCoIndHyp") <$> mrFunBody f2 args2
+     (pre1, pre2) <- applyCoIndHypPreconds hyp
+     pre <- liftSC2 scAnd pre1 pre2
+     (withOnlyUVars (coIndHypCtx hyp) $ withOnlyAssumption pre $
```

Comment:
I feel like we loose some generality here by getting rid of all our other assumptions. Like, what if we have nested co-inductions and we need a precondition from the outer co-induction to prove something in the inner? Though we don't have any examples like that as the moment, so it's probably fine for now.

Reply:
Actually, the `withOnlyUVars` is going to throw away any existing assumptions anyway, so technically the `withOnlyAssumption` is superfluous, but it's helpful to have the "Only" there for documentation